File: packages/server/src/services/github-oauth.service.ts
import { randomBytes } from 'crypto';
import { createLogger } from '../utils/logger.js';
import type { GitHubAuthUrlResponse, GitHubCallbackResponse } from '@dash-agent/shared';
import { saveSecret } from './secrets.service.js';
import { resetGitHubClient } from '../github/client.js';

const logger = createLogger('services:github-oauth');

/**
 * GitHub OAuth configuration
 * These would typically come from environment variables
 */
const GITHUB_CLIENT_ID = process.env['GITHUB_CLIENT_ID'] || '';
const GITHUB_CLIENT_SECRET = process.env['GITHUB_CLIENT_SECRET'] || '';
const GITHUB_REDIRECT_URI = process.env['GITHUB_REDIRECT_URI'] || 'http://localhost:3003/setup/github/callback';

/**
 * In-memory store for OAuth state tokens
 * In production, this should use Redis or a database
 */
const stateTokens = new Map<string, { createdAt: Date; expiresAt: Date }>();

/**
 * Clean up expired state tokens periodically
 */
setInterval(() => {
  const now = new Date();
  for (const [token, data] of stateTokens.entries()) {
    if (data.expiresAt < now) {
      stateTokens.delete(token);
    }
  }
}, 60 * 1000); // Clean up every minute

/**
 * Generates a secure random state token for OAuth
 */
function generateStateToken(): string {
  return randomBytes(32).toString('hex');
}

/**
 * Generates the GitHub OAuth authorization URL.
 *
 * @returns Object containing the auth URL and state token
 */
export function generateAuthUrl(): GitHubAuthUrlResponse {
  logger.debug('Generating GitHub OAuth URL');

  const state = generateStateToken();

  // Store state token with 10 minute expiration
  const now = new Date();
  stateTokens.set(state, {
    createdAt: now,
    expiresAt: new Date(now.getTime() + 10 * 60 * 1000),
  });

  // Required scopes for dash-agent:
  // - repo: Full control of private repositories (needed for PR creation)
  // - read:user: Read user profile data
  const scopes = ['repo', 'read:user'];

  const params = new URLSearchParams({
    client_id: GITHUB_CLIENT_ID,
    redirect_uri: GITHUB_REDIRECT_URI,
    scope: scopes.join(' '),
    state,
  });

  const authUrl = `https://github.com/login/oauth/authorize?${params.toString()}`;

  logger.info('Generated GitHub OAuth URL', { state });

  return { authUrl, state };
}

/**
 * Validates a state token from the OAuth callback.
 *
 * @param state - The state token to validate
 * @returns True if the state is valid, false otherwise
 */
export function validateState(state: string): boolean {
  const tokenData = stateTokens.get(state);

  if (!tokenData) {
    logger.warn('OAuth state not found', { state });
    return false;
  }

  const now = new Date();
  if (tokenData.expiresAt < now) {
    logger.warn('OAuth state expired', { state });
    stateTokens.delete(state);
    return false;
  }

  // State is valid, remove it (single use)
  stateTokens.delete(state);
  return true;
}

/**
 * Exchanges an authorization code for an access token.
 *
 * @param code - The authorization code from GitHub
 * @returns The access token or null if exchange failed
 */
async function exchangeCodeForToken(code: string): Promise<string | null> {
  logger.debug('Exchanging code for access token');

  try {
    const response = await fetch('https://github.com/login/oauth/access_token', {
      method: 'POST',
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        client_id: GITHUB_CLIENT_ID,
        client_secret: GITHUB_CLIENT_SECRET,
        code,
      }),
    });

    if (!response.ok) {
      logger.error('Token exchange failed', { status: response.status });
      return null;
    }

    const data = await response.json() as {
      access_token?: string;
      error?: string;
      error_description?: string;
    };

    if (data.error) {
      logger.error('Token exchange error', {
        error: data.error,
        description: data.error_description
      });
      return null;
    }

    return data.access_token || null;
  } catch (error) {
    logger.error('Token exchange error', {
      error: error instanceof Error ? error.message : String(error)
    });
    return null;
  }
}

/**
 * Fetches GitHub user information using an access token.
 *
 * @param token - The GitHub access token
 * @returns User information or null if fetch failed
 */
async function fetchUserInfo(token: string): Promise<{ login: string; avatar_url: string } | null> {
  logger.debug('Fetching GitHub user info');

  try {
    const response = await fetch('https://api.github.com/user', {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Accept': 'application/vnd.github.v3+json',
        'User-Agent': 'dash-agent',
      },
    });

    if (!response.ok) {
      logger.error('User info fetch failed', { status: response.status });
      return null;
    }

    const data = await response.json() as { login: string; avatar_url: string };
    return { login: data.login, avatar_url: data.avatar_url };
  } catch (error) {
    logger.error('User info fetch error', {
      error: error instanceof Error ? error.message : String(error)
    });
    return null;
  }
}

/**
 * Handles the OAuth callback by exchanging the code for a token,
 * storing it encrypted in the server, and fetching user information.
 *
 * @param code - The authorization code from GitHub
 * @param state - The state token to validate
 * @param storeToken - If true, stores the token server-side (default: true)
 * @returns Object containing success status and user info (token NOT returned)
 */
export async function handleCallback(
  code: string,
  state: string,
  storeToken: boolean = true
): Promise<GitHubCallbackResponse> {
  logger.info('Handling GitHub OAuth callback');

  // Validate state
  if (!validateState(state)) {
    return {
      success: false,
      error: 'Invalid or expired state token. Please try again.',
    };
  }

  // Check if GitHub OAuth is configured
  if (!GITHUB_CLIENT_ID || !GITHUB_CLIENT_SECRET) {
    logger.warn('GitHub OAuth not configured, simulating success');
    // For development without OAuth credentials, simulate success
    // In dev mode, we still save a simulated token for testing
    if (storeToken) {
      const metadata: Record<string, unknown> = {
        username: 'demo-user',
        avatarUrl: 'https://github.com/identicons/demo.png',
        connectionMethod: 'oauth',
      };
      saveSecret('github_token', 'github', 'simulated-token-for-development', metadata);
      resetGitHubClient();
    }
    return {
      success: true,
      username: 'demo-user',
      avatarUrl: 'https://github.com/identicons/demo.png',
    };
  }

  // Exchange code for token
  const token = await exchangeCodeForToken(code);
  if (!token) {
    return {
      success: false,
      error: 'Failed to exchange authorization code for access token.',
    };
  }

  // Fetch user info
  const userInfo = await fetchUserInfo(token);
  if (!userInfo) {
    return {
      success: false,
      error: 'Failed to fetch user information from GitHub.',
    };
  }

  // Store the token encrypted in the server
  if (storeToken) {
    const metadata: Record<string, unknown> = {
      username: userInfo.login,
      avatarUrl: userInfo.avatar_url,
      connectionMethod: 'oauth',
    };
    saveSecret('github_token', 'github', token, metadata);
    resetGitHubClient(); // Reset client so it picks up the new token
    logger.info('GitHub token stored successfully via OAuth', { username: userInfo.login });
  }

  logger.info('GitHub OAuth completed successfully', { username: userInfo.login });

  // Note: Token is NOT returned to frontend - it's stored server-side
  return {
    success: true,
    username: userInfo.login,
    avatarUrl: userInfo.avatar_url,
  };
}

/**
 * Validates a GitHub token by making a test API call.
 *
 * @param token - The GitHub token to validate
 * @returns True if the token is valid, false otherwise
 */
export async function validateToken(token: string): Promise<boolean> {
  logger.debug('Validating GitHub token');

  try {
    const response = await fetch('https://api.github.com/user', {
      headers: {
        'Authorization': `Bearer ${token}`,
        'Accept': 'application/vnd.github.v3+json',
        'User-Agent': 'dash-agent',
      },
    });

    return response.ok;
  } catch (error) {
    logger.error('Token validation error', {
      error: error instanceof Error ? error.message : String(error)
    });
    return false;
  }
}

/**
 * Checks if GitHub OAuth is configured.
 *
 * @returns True if client ID and secret are set
 */
export function isOAuthConfigured(): boolean {
  return Boolean(GITHUB_CLIENT_ID && GITHUB_CLIENT_SECRET);
}

File: packages/server-rs/src/services/github_oauth_service.rs
//! GitHub OAuth service for managing the OAuth2 authorization flow.
//!
//! Port of `packages/server/src/services/github-oauth.service.ts`.
//!
//! Manages:
//! - Generation of OAuth authorization URLs with CSRF state tokens
//! - Token exchange (authorization code -> access token)
//! - In-memory state token store with expiration

use std::collections::HashMap;
use std::sync::RwLock;
use std::time::Instant;

use serde::{Deserialize, Serialize};
use tracing::{debug, error, info, warn};

use crate::error::AppError;

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/// Response containing the OAuth authorization URL.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuthUrlResponse {
    pub auth_url: String,
    pub state: String,
}

/// Response from the OAuth callback.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CallbackResponse {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub username: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub avatar_url: Option<String>,
}

/// Token returned from the exchange endpoint.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct OAuthToken {
    pub access_token: String,
    pub username: String,
    pub avatar_url: String,
}

/// Internal state entry.
struct StateEntry {
    #[allow(dead_code)]
    created_at: Instant,
    expires_at: Instant,
}

// ---------------------------------------------------------------------------
// GitHub OAuth Service
// ---------------------------------------------------------------------------

/// Service for managing GitHub OAuth2 authorization flow.
pub struct GitHubOAuthService {
    client_id: Option<String>,
    client_secret: Option<String>,
    redirect_uri: String,
    /// In-memory store for OAuth state tokens (state -> expiry).
    pending_states: RwLock<HashMap<String, StateEntry>>,
}

impl GitHubOAuthService {
    /// Creates a new service, reading configuration from environment variables.
    pub fn new() -> Self {
        let client_id = std::env::var("GITHUB_CLIENT_ID").ok().filter(|s| !s.is_empty());
        let client_secret = std::env::var("GITHUB_CLIENT_SECRET").ok().filter(|s| !s.is_empty());
        let redirect_uri = std::env::var("GITHUB_REDIRECT_URI")
            .unwrap_or_else(|_| "http://localhost:3003/setup/github/callback".into());

        Self {
            client_id,
            client_secret,
            redirect_uri,
            pending_states: RwLock::new(HashMap::new()),
        }
    }

    /// Returns `true` if GitHub OAuth is configured (client ID and secret are set).
    pub fn is_configured(&self) -> bool {
        self.client_id.is_some() && self.client_secret.is_some()
    }

    /// Generates a GitHub OAuth authorization URL.
    ///
    /// Returns the URL to redirect the user to and the CSRF state token.
    pub fn get_auth_url(&self) -> Result<AuthUrlResponse, AppError> {
        debug!("Generating GitHub OAuth URL");

        let state = Self::generate_state_token();

        // Store state with 10 minute expiration
        {
            let now = Instant::now();
            let mut states = self
                .pending_states
                .write()
                .map_err(|_| AppError::Internal(anyhow::anyhow!("State lock poisoned")))?;
            states.insert(
                state.clone(),
                StateEntry {
                    created_at: now,
                    expires_at: now + std::time::Duration::from_secs(600),
                },
            );
        }

        let client_id = self.client_id.as_deref().unwrap_or("");
        let scopes = "repo read:user";

        let auth_url = format!(
            "https://github.com/login/oauth/authorize?client_id={}&redirect_uri={}&scope={}&state={}",
            urlencoding::encode(client_id),
            urlencoding::encode(&self.redirect_uri),
            urlencoding::encode(scopes),
            urlencoding::encode(&state),
        );

        info!(state = %state, "Generated GitHub OAuth URL");
        Ok(AuthUrlResponse { auth_url, state })
    }

    /// Validates and consumes a state token from the OAuth callback.
    pub fn validate_state(&self, state: &str) -> bool {
        let mut states = match self.pending_states.write() {
            Ok(s) => s,
            Err(_) => return false,
        };

        match states.remove(state) {
            Some(entry) => {
                if Instant::now() > entry.expires_at {
                    warn!(state, "OAuth state expired");
                    false
                } else {
                    true
                }
            }
            None => {
                warn!(state, "OAuth state not found");
                false
            }
        }
    }

    /// Exchanges an authorization code for an access token and fetches user info.
    pub async fn exchange_code(
        &self,
        code: &str,
        state: &str,
    ) -> Result<OAuthToken, AppError> {
        info!("Handling GitHub OAuth callback");

        // Validate state
        if !self.validate_state(state) {
            return Err(AppError::Validation(
                "Invalid or expired state token. Please try again.".into(),
            ));
        }

        // Check configuration
        if !self.is_configured() {
            warn!("GitHub OAuth not configured, returning simulated response");
            return Ok(OAuthToken {
                access_token: "simulated-token-for-development".into(),
                username: "demo-user".into(),
                avatar_url: "https://github.com/identicons/demo.png".into(),
            });
        }

        let client_id = self.client_id.as_deref().unwrap_or("");
        let client_secret = self.client_secret.as_deref().unwrap_or("");

        // Exchange code for token
        let client = reqwest::Client::new();
        let response = client
            .post("https://github.com/login/oauth/access_token")
            .header("Accept", "application/json")
            .header("Content-Type", "application/json")
            .json(&serde_json::json!({
                "client_id": client_id,
                "client_secret": client_secret,
                "code": code,
            }))
            .send()
            .await
            .map_err(|e| AppError::Internal(anyhow::anyhow!("Token exchange request failed: {e}")))?;

        if !response.status().is_success() {
            error!(status = %response.status(), "Token exchange failed");
            return Err(AppError::Internal(anyhow::anyhow!("Token exchange failed")));
        }

        #[derive(Deserialize)]
        struct TokenResponse {
            access_token: Option<String>,
            error: Option<String>,
            error_description: Option<String>,
        }

        let token_data: TokenResponse = response
            .json()
            .await
            .map_err(|e| AppError::Internal(anyhow::anyhow!("Failed to parse token response: {e}")))?;

        if let Some(ref err) = token_data.error {
            error!(
                error = %err,
                description = token_data.error_description.as_deref().unwrap_or(""),
                "Token exchange error"
            );
            return Err(AppError::Internal(anyhow::anyhow!(
                "Token exchange error: {}",
                token_data.error_description.as_deref().unwrap_or(err)
            )));
        }

        let access_token = token_data.access_token.ok_or_else(|| {
            AppError::Internal(anyhow::anyhow!("No access token in response"))
        })?;

        // Fetch user info
        let user_response = client
            .get("https://api.github.com/user")
            .header("Authorization", format!("Bearer {access_token}"))
            .header("Accept", "application/vnd.github.v3+json")
            .header("User-Agent", "agent-board")
            .send()
            .await
            .map_err(|e| AppError::Internal(anyhow::anyhow!("User info request failed: {e}")))?;

        if !user_response.status().is_success() {
            error!("Failed to fetch user info after token exchange");
            return Err(AppError::Internal(anyhow::anyhow!(
                "Failed to fetch user information from GitHub"
            )));
        }

        #[derive(Deserialize)]
        struct UserInfo {
            login: String,
            avatar_url: String,
        }

        let user: UserInfo = user_response
            .json()
            .await
            .map_err(|e| AppError::Internal(anyhow::anyhow!("Failed to parse user info: {e}")))?;

        info!(username = %user.login, "GitHub OAuth completed successfully");

        Ok(OAuthToken {
            access_token,
            username: user.login,
            avatar_url: user.avatar_url,
        })
    }

    /// Cleans up expired state tokens. Call periodically (e.g., every 60s).
    pub fn cleanup_expired_states(&self) {
        if let Ok(mut states) = self.pending_states.write() {
            let now = Instant::now();
            states.retain(|_, entry| entry.expires_at > now);
        }
    }

    /// Generates a cryptographically secure random state token (64 hex chars).
    fn generate_state_token() -> String {
        use rand::Rng;
        let mut rng = rand::thread_rng();
        let bytes: [u8; 32] = rng.gen();
        hex::encode(bytes)
    }
}

impl Default for GitHubOAuthService {
    fn default() -> Self {
        Self::new()
    }
}

/// Validates a GitHub token by making a test API call.
pub async fn validate_token(token: &str) -> bool {
    let client = match reqwest::Client::builder()
        .default_headers({
            let mut h = reqwest::header::HeaderMap::new();
            h.insert("User-Agent", "agent-board".parse().unwrap());
            h
        })
        .build()
    {
        Ok(c) => c,
        Err(_) => return false,
    };

    match client
        .get("https://api.github.com/user")
        .header("Authorization", format!("Bearer {token}"))
        .header("Accept", "application/vnd.github.v3+json")
        .send()
        .await
    {
        Ok(resp) => resp.status().is_success(),
        Err(_) => false,
    }
}
